smdebug/core/tfevent/timeline_writer.py

```diff
@@ -0,0 +1,203 @@
+# Standard Library
+import collections
+import json
+import time
+
+# First Party
+from smdebug.core.access_layer.file import TSAccessFile
+from smdebug.core.access_layer.s3 import TSAccessS3
+from smdebug.core.logger import get_logger
+from smdebug.core.utils import is_s3
+
+
+"""
+TimelineRecord represents one trace event that ill be written into a trace event JSON file.
+"""
+
+
+class TimelineRecord:
+    def __init__(
+        self, training_phase="", phase="X", operator_name="", args=None, timestamp=0, duration=0
+    ):
+        """
+
+        :param training_phase: strings like, data_iterating, forward, backward, operations etc
+        :param phase: Trace event phases. Example "X", "B", "E", "M"
+        :param operator_name: more details about phase like whether dataset or iterator
+        :param args: other information to be added as args
+        :param timestamp: start_time for the event
+        :param duration: any duration manually computed (in seconds)
+        """
+        self.training_phase = training_phase
+        self.phase = phase
+        self.op_name = operator_name
+        self.args = args
+        self.ts_micros = (
+            int(timestamp * 1000000) if timestamp else int(round(time.time() * 1000000))
+        )
+        self.duration = duration
```

Comment:
you should check if event is 'X' and duration is not provided then calculate duration here. by doing now() - event_ts